pytorch_lightning/core/datamodule.py

```diff
@@ -366,15 +366,25 @@ def wrapped_fn(*args: str, **kwargs: Optional[str]) -> Any:
                 stage = args[0] if len(args) else kwargs.get("stage", None)
 
                 if stage is None:
+                    has_run = True
                     for s in ("fit", "validate", "test"):
-                        setattr(obj, f"_has_{name}_{s}", True)
+                        attr = f"_has_{name}_{s}"
+                        has_run &= getattr(obj, attr)
+                        setattr(obj, attr, True)
                 else:
-                    setattr(obj, f"_has_{name}_{stage}", True)
+                    attr = f"_has_{name}_{stage}"
+                    has_run = getattr(obj, attr)
+                    setattr(obj, attr, True)
 
             elif name == "prepare_data":
+                has_run = obj._has_prepared_data
                 obj._has_prepared_data = True
 
-            return fn(*args, **kwargs)
+            else:
+                raise ValueError(name)
```

Comment:
I believe this is unreachable. See the test I added in the last commit. We get AttributeError before that.

Reply:
I removed the `ValueError`, it'd only appear if you were to do `dm.sunflower("fit")` but that would not work as we manually wrap these functions so it's pointless to have.